Initial version with docker container and docker-compose

oc013 opened the pull request:
* Cache large model files
* Use ttf included in file for prompt matrix
* Run all unbuffered for output to stdout

Requires docker-compose 1.28+ and nvidia-container-toolkit for gpu pass through

Miniconda etc handled.

Launching simple as:
```
git clone <repourl>
docker-compose up
```

Tested on Ubuntu 20.04

hlky commented on entrypoint.sh:
```diff
@@ -0,0 +1,35 @@
+conda activate ldm
+
+# Validate model files
+echo "Validating model files..."
```
you can add esrgan models now too it's enabled in environment.yaml

oc013 replied:
I'll try to wrap that all up into a function and pre-download everything. One thing to note is for some of these additional files they will be cached in the `conda_env` volume, so may not even need this additional code/verification to occur.  (I did notice the `GFPGAN` model downloaded automatically when I had initially placed it in the wrong path during testing)

hlky replied:
nice, let me know when done and i can merge, good work

oc013 replied:
I made some updates which should make things reasonably maintainable and also added the ESRGAN models you mentioned. Adding a more in depth comment below of changes.

hlky commented on scripts/relauncher.py:
```diff
@@ -5,7 +5,7 @@
     if n > 0:
         print(f'\tRelaunch count: {n}')
-    os.system("python scripts/webui.py")
+    os.system("python -u scripts/webui.py")
```
is unbuffered output a quirk of docker support? i can't think of any reason it will affect regular users, but i will wait to confirm before i merge

oc013 replied:
Very possible that is the case. Without setting unbuffered I was not even seeing `Relauncher: Launching...` or any other messages in STDOUT until after I had loaded the UI and submitted something

oc013 replied:
Hopefully you were able to confirm, I do not have a windows system to test.

hlky reviewed (changes requested):
docker will be useful for some and won't affect existing users in any way
good work, thanks!

oc013 commented:
In `entrypoint.sh` I added an array at the top which contains all model files to be downloaded that are not included in this repo.  It's space delimited, if there's every a case where you need spaces in any of the columns I believe you can place it in double quotes. 

Each line should look like:
```
'<local filename> <local path> <remote url> <sha256sum>'
```
To the docker container, the local path always starts with `/src` which is the root of the repo, hence the `/src/src/` you see in there. Later there's a for loop and the function `validateDownloadModel()` that consumes this data, validates the hash and performs the download if needed.

I also added the automatic conda env update like you added in the windows script. It didn't seem to recognize the existing downloads of the `pip -e git` dependencies and was recreating/redownloading those every run. I made a simple caching system with unix timestamps that should do the trick based on when the docker image was first created and/or when `environment.yaml` was last modified.

Also added another volume to cache the root user's home directory in the container to prevent certain pip dependencies from redownloading data every time the container was restarted.

Overall I think it's working pretty solidly, the sha256sum is a bit slow on the model.ckpt file but I think it's worth running every time in case the download is interrupted or it gets corrupted somehow.
